Tests/GRPCHTTP2TransportTests/HTTP2TransportNIOPosixTests.swift

```diff
@@ -0,0 +1,166 @@
+import GRPCHTTP2Core
+import GRPCHTTP2TransportNIOPosix
+import XCTest
+
+@available(macOS 14.0, iOS 17.0, watchOS 10.0, tvOS 17.0, *)
+final class HTTP2TransportNIOPosixTests: XCTestCase {
+  func testGetListeningAddress_IPv4() async throws {
+    let transport = GRPCHTTP2Core.HTTP2ServerTransport.Posix(
+      address: .ipv4(host: "0.0.0.0", port: 0)
+    )
+
+    try await withThrowingDiscardingTaskGroup { group in
+      group.addTask {
+        try await transport.listen { _ in }
+      }
+
+      group.addTask {
+        let address = try await transport.listeningAddress
+        XCTAssertNotNil(address.ipv4)
+        XCTAssertNotEqual(address.ipv4!.port, 0)
+        transport.stopListening()
+      }
+    }
+  }
+
+  func testGetListeningAddress_IPv6() async throws {
+    let transport = GRPCHTTP2Core.HTTP2ServerTransport.Posix(address: .ipv6(host: "::1", port: 0))
+
+    try await withThrowingDiscardingTaskGroup { group in
+      group.addTask {
+        try await transport.listen { _ in }
+      }
+
+      group.addTask {
+        let address = try await transport.listeningAddress
+        XCTAssertNotNil(address.ipv6)
+        XCTAssertNotEqual(address.ipv6!.port, 0)
+        transport.stopListening()
+      }
+    }
+  }
+
+  func testGetListeningAddress_UnixDomainSocket() async throws {
+    let transport = GRPCHTTP2Core.HTTP2ServerTransport.Posix(
+      address: .unixDomainSocket(path: "/tmp/test")
+    )
+
+    try await withThrowingDiscardingTaskGroup { group in
+      group.addTask {
+        try await transport.listen { _ in }
+      }
+
+      group.addTask {
+        let address = try await transport.listeningAddress
+        XCTAssertEqual(
+          address.unixDomainSocket,
+          GRPCHTTP2Core.SocketAddress.UnixDomainSocket(path: "/tmp/test")
+        )
+        transport.stopListening()
+      }
+    }
+  }
+
+  func testGetListeningAddress_Vsock() async throws {
+    try XCTSkipUnless(self.vsockAvailable(), "Vsock unavailable")
+
+    let transport = GRPCHTTP2Core.HTTP2ServerTransport.Posix(
+      address: .vsock(contextID: .any, port: .any)
+    )
+
+    try await withThrowingDiscardingTaskGroup { group in
+      group.addTask {
+        try await transport.listen { _ in }
+      }
+
+      group.addTask {
+        let address = try await transport.listeningAddress
+        XCTAssertNotNil(address.virtualSocket)
+        transport.stopListening()
+      }
+    }
+  }
+
+  func testGetListeningAddress_InvalidAddress() async {
+    let transport = GRPCHTTP2Core.HTTP2ServerTransport.Posix(
+      address: .unixDomainSocket(path: "")
+    )
+    let errorThrown = _LockedValueBox(false)
+
+    try? await withThrowingDiscardingTaskGroup { group in
+      group.addTask {
+        try await transport.listen { _ in }
+      }
+
+      group.addTask {
+        do {
+          _ = try await transport.listeningAddress
+        } catch let error as RuntimeError {
+          errorThrown.withLockedValue { $0 = true }
```

Comment:
We don't need a locked value box here, if we `XCTFail` on the happy-path then we know each branch is covered. We also know that all must run and complete before the task group returns.